crate/operator/restore_backup.py

```diff
@@ -0,0 +1,925 @@
+# additional information regarding copyright ownership.  Crate licenses
+# this file to you under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.  You may
+# obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+# However, if you have executed another commercial license agreement
+# with Crate these terms will supersede the license and you may use the
+# software solely pursuant to the terms of the relevant commercial agreement.
+
+import asyncio
+import logging
+import re
+from typing import Any, List, Optional
+
+import kopf
+from aiohttp.client_exceptions import WSServerHandshakeError
+from aiopg import Cursor
+from kubernetes_asyncio.client import ApiException, CoreV1Api, CustomObjectsApi
+from kubernetes_asyncio.client.api_client import ApiClient
+from kubernetes_asyncio.stream import WsApiClient
+from psycopg2 import DatabaseError, OperationalError, ProgrammingError
+from psycopg2.errors import DuplicateTable
+from psycopg2.extensions import AsIs, QuotedString, quote_ident
+
+from crate.operator.config import config
+from crate.operator.constants import API_GROUP, RESOURCE_CRATEDB, SYSTEM_USERNAME
+from crate.operator.cratedb import (
+    connection_factory,
+    get_cluster_settings,
+    set_cluster_setting,
+)
+from crate.operator.operations import (
+    get_cratedb_resource,
+    scale_backup_metrics_deployment,
+)
+from crate.operator.utils import crate
+from crate.operator.utils.kopf import StateBasedSubHandler, subhandler_partial
+from crate.operator.utils.kubeapi import (
+    get_host,
+    get_system_user_password,
+    resolve_secret_key_ref,
+)
+from crate.operator.utils.notifications import send_operation_progress_notification
+from crate.operator.webhooks import (
+    WebhookEvent,
+    WebhookFeedbackPayload,
+    WebhookOperation,
+    WebhookStatus,
+)
+
+RESTORE_BACKUP_SECRETS: List[str] = [
+    "bucket",
+    "secretAccessKey",
+    "basePath",
+    "accessKeyId",
+]
+RESTORE_MAX_BYTES_PER_SEC: str = "200mb"
+RESTORE_CLUSTER_CONCURRENT_REBALANCE: int = 6
+DEFAULT_MAX_BYTES_PER_SEC: str = "40mb"
+DEFAULT_CLUSTER_CONCURRENT_REBALANCE: int = 2
+CRASH_COMMAND_DELAY: int = 30
+
+
+def is_valid_snapshot(new: kopf.Body, **kwargs) -> bool:
+    """
+    This checks if the new snapshot name is valid or not (empty).
+    This check is necessary to avoid another restore operation is
+    triggered after we reset the snapshot field at the end of a
+    completed restore operation.
+
+    :param new: The new CrateDB resource
+    """
+    return len(new["spec"]["cluster"]["restoreSnapshot"]["snapshot"]) > 0
```

Comment:
Hm true, if the whole `restoreSnapshot` object was deleted, it raises an error, will fix it. Only `snapshot` cannot be `None` I think since it is defined as `required` and `type: string`